internal/provider/base_eda_datasource.go

```diff
@@ -0,0 +1,198 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/datasourcevalidator"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	tfpath "github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource                     = &BaseEdaDataSource{}
+	_ datasource.DataSourceWithConfigure        = &BaseEdaDataSource{}
+	_ datasource.DataSourceWithConfigValidators = &BaseEdaDataSource{}
+	_ datasource.DataSourceWithValidateConfig   = &BaseEdaDataSource{}
+)
+
+// NewBaseEdaDataSource creates a new `BaseEdaDataSource`.
+func NewBaseEdaDataSource(client ProviderHTTPClient, stringDescriptions StringDescriptions) *BaseEdaDataSource {
+	return &BaseEdaDataSource{
+		client:             client,
+		StringDescriptions: stringDescriptions,
+	}
+}
+
+// Metadata returns the data source type name composing it from the provider type name and the
+// entity slug string passed in the constructor.
+func (d *BaseEdaDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_%s", req.ProviderTypeName, d.MetadataEntitySlug)
+	fmt.Println(resp.TypeName)
+}
+
+// Schema defines the schema fields for the data source.
+func (d *BaseEdaDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"name": schema.StringAttribute{
+				Required: true,
+			},
+			"url": schema.StringAttribute{
+				Computed: true,
+			},
+			"id": schema.Int64Attribute{
+				Computed: true,
+			},
+		},
+	}
+}
```

Comment:
Since `func Schema()` is implemented here in `BaseEdaDataSource` (and not on the concrete `EDAEventStreamDataSource` struct), that suggests we expect every `datasource` that we will implement under EDA to have these attributes `name`, `url`, and `id`. Is that the intent?

Reply:
I had looked at a few of the object definitions in the Swagger docs and the ones I looked at (Resource, ResourceType, RoleDefinition) all have those attributes. Looking a second time at other objects (ActivationInstance and Rulebook) they have the `id` and `name`, but not the `url`.

I think this method can be moved into the `EDAEventStreamDataSource` definition as you mention.

Reply:
I've switched to using the same pattern used by the BaseResource struct, where a `GetBaseAttributes` function allows descendant structs to use the base attributes and add its own, while still providing a base implementation of the `Schema` function.